bve-build/src/shaders.rs

```diff
@@ -0,0 +1,136 @@
+use itertools::Itertools;
+use nom::{
+    branch::alt,
+    bytes::complete::{tag, tag_no_case, take_while},
+    combinator::opt,
+    multi::separated_list,
+    sequence::{delimited, preceded, tuple},
+    IResult, InputTakeAtPosition,
+};
+
+#[derive(Debug, Copy, Clone)]
+pub enum ShaderType {
+    Vertex,
+    Fragment,
+    Compute,
+}
+
+#[derive(Debug, Clone)]
+pub struct ShaderCombination<'a> {
+    pub name: &'a str,
+    pub ty: ShaderType,
+    pub defines: Vec<SingleDefine<'a>>,
+}
+
+pub fn parse_shader_compile_file(input: &str) -> Result<impl Iterator<Item = ShaderCombination<'_>>, &str> {
+    separated_list(alt((tag("\r\n"), tag("\n"))), parse_shader_directive)(input)
+        .map_err(|_| "parsing error")
+        .and_then(|(str, v)| {
+            if !str.is_empty() {
+                return Err(str);
+            }
+            Ok(v.into_iter().flatten())
+        })
+}
+
+/// Takes a parser and wraps it so it delimited with whitespace
+fn w<F, I, O>(func: F) -> impl Fn(I) -> IResult<I, O>
```

Comment:
Rename

Reply:
This is the one time I'm going to say no, as this function gets spammed a lot in parsers, and the name needs to be as short as possible to let you still see the structure of the parser.